pkg/extensions/tbtc/tbtc.go

```diff
@@ -97,10 +119,296 @@ func (t *tbtc) monitorRetrievePubKey(
 	return nil
 }
 
+func (t *tbtc) monitorProvideRedemptionSignature(
+	ctx context.Context,
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) error {
+	monitoringStartFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		// Start right after a redemption has been requested or the redemption
+		// fee has been increased.
+		return t.chain.OnDepositRedemptionRequested(handler)
+	}
+
+	monitoringStopFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		// Stop in case the redemption signature has been provided by someone else.
+		signatureSubscription, err := t.chain.OnDepositGotRedemptionSignature(handler)
+		if err != nil {
+			return nil, err
+		}
+
+		// Stop in case the redemption proof has been provided by someone else.
+		redeemedSubscription, err := t.chain.OnDepositRedeemed(handler)
+		if err != nil {
+			return nil, err
+		}
+
+		return subscription.NewEventSubscription(
+			func() {
+				signatureSubscription.Unsubscribe()
+				redeemedSubscription.Unsubscribe()
+			},
+		), nil
+	}
+
+	actFn := func(depositAddress string) error {
+		keepAddress, err := t.chain.KeepAddress(depositAddress)
+		if err != nil {
+			return err
+		}
+
+		redemptionRequestedEvents, err := t.chain.PastDepositRedemptionRequestedEvents(
+			depositAddress,
+			t.pastEventsLookupStartBlock(),
+		)
+		if err != nil {
+			return err
+		}
+
+		if len(redemptionRequestedEvents) == 0 {
+			return fmt.Errorf(
+				"no redemption requested events found for deposit: [%v]",
+				depositAddress,
+			)
+		}
+
+		latestRedemptionRequestedEvent :=
+			redemptionRequestedEvents[len(redemptionRequestedEvents)-1]
+
+		signatureSubmittedEvents, err := t.chain.PastSignatureSubmittedEvents(
+			keepAddress,
+			latestRedemptionRequestedEvent.BlockNumber,
+		)
+		if err != nil {
+			return err
+		}
+
+		depositDigest := latestRedemptionRequestedEvent.Digest
+
+		// Start iterating from the latest event.
+		for i := len(signatureSubmittedEvents) - 1; i >= 0; i-- {
```

Comment:
Yep now we force the order. But, I'd leave the reversed loop just in case.

Reply:
I think it's misleading. If we are leaving a loop here because there is an edge case where we can not just take the last event, we should also do something with `redemptionRequestedEvents`. Happy to be proven wrong but I do not think there is such an edge case for tBTC extension.

Reply:
Ok, I'll remove that loop.